### ISSUE TITLE
New pages for project coordinators

### ISSUE THREAD

**Issue by fosterlynn** [FreedomCoop/valuenetwork#132]
I'm starting a new issue for this topic.  This is coming from: https://github.com/FreedomCoop/valuenetwork/issues/131#issuecomment-243110629 and includes the following:

>
* design and code the sub-pages of 'your projects' page for coordinators (managers): contributions, active orders, exchanges, etc... see #123 and #108
* create the coordinators workflow for plan process tasks in a project: huge subject to design and work-on from the member's point of view, and include all needed pages with restricted scope in the work app.

Next I'll put together an initial look at what might be included, and possible nav/workflow, for discussion on priorities.

**Comment by fosterlynn** [comment]
The following is what a project coordinator might want to do, for discussion.  All assumed to be filtered by the projects the person is coordinator for, with issues noted.  Current page reference listed after each item in brackets.  A lot of these will be lower priority, but we'll need to discuss to figure out the priority order.

Orders and plans
* View open customer orders and work orders [demand]
  * View individual order [order-schedule]
  * View closed orders [closed-work-orders, we don't have one for customer orders now]
* Create customer order (which will create a plan if there are recipes) [create-order]
* Create work order & plan without recipe [process-selections]
* Create work order & plan with recipe [plan-from-recipe]
* Change order plan (same as "View individual order" above) [order-schedule]
  * Includes adding a deliverable, planning forward, changing dates, adding processes (workflow only)
* Open purchase needs [supply]
  * Note these are when you plan something and there is not a known way in the system to make something you need
  * There is also a list by possible supplier, which may or may not work right now

Exchanges
* Create or edit an exchange [exchange]
  * Note right now getting to this page to create an exchange is from different places based on incoming, internal, outgoing.  Incoming is for purchases into the network, outgoing is for sales and other transfers out of the network, and internal is inside.  I'm not sure I like those divisions any more.  But in any case, you need a place to jump on to the exchange page after picking an exchange type.
  * Note that getting to this page to edit an exchange is from the exchange lists.
  * Note that an outgoing (sale type) exchange is automatically created when you create a customer order.
* Exchange lists [demand-exchanges, exchanges (supply only), internal-exchanges]
  * Note these could be combined into one, right now are linked from Demand page, Supply page, and Organization page.

Distribution of income
* List of value equations [value-equations]
* Create and edit value equation [edit-value-equation]
* Value equation sandbox, to test value equations [value-equation-sandbox]
* Make a distribution [create-distribution-using-value-equation]
* Distribution list [distributions]
* Edit a distribution [distribution]

Resource types and recipes
* Note this is a bigger issue, because resource types are currently not associated with context agents / projects.  See https://github.com/FreedomCoop/valuenetwork/issues/130. 
* Resource types list [resources]
  * Create resource type [this is a button/popup from the resource types page]
  * Edit/delete resource type [resource-type]
  * Create/edit recipe for the resource type [edit-xbomfg]
  * View recipe for the resource type [xbomfg]
* Resource type lists (combined recipes to plan together) [resource-type-lists]
  * Create list [create-resource-type-list]
  * Change list [change-resource-type-list]

Resources / inventory
* Note we can more easily make resources belong to a context agent, through the existing AgentResourceRole.  Would need to set up a type that we know means the resource belongs in the context of a project.  Or we could possibly depend on the history of the resource to tell.
* Resource list (Inventory) [inventory]
  * Create resource [currently a button on the Resource Type page, only for beginning balances]
  * Edit resource [resource]
  * Event history for a resource (report) [event-history]
  * Incoming value flow (report) [incoming-value-flows]
  * Adjust quantity (to correct inventory) [button on the event history now]
* Locations for resources [locations]
  * Create location [create-location]

Other related to current work (or maybe these should be for non-coordinators too?)
* Work in process report (project-wip)
* Finished processes report (finished-processes)
* Contributions report [contributions]
* Work notes history (for projects that keep detailed process notes) [labnotes-history]

Then there are a bunch of questions around other setup.  These include process patterns, exchange types, agent association types, units, facets and maybe more.  I think we should limit those to admins right now.  Project coordinators can ask admins to help them with these things if they need something new.

**Comment by fosterlynn** [comment]
Thinking about how to group these in terms of user stories or use cases, so that we can implement minimal features that fit together.  The following might a couple first steps.

Minimal work planning
  * Implement process logging for all members
  * View open work orders
  * Create work order & plan without recipe
  * Change order plan

Then can add customer orders and/or recipes and/or project reporting later.  And see if this can be done without them needing to create their own resource types. 

Exchanges (Decentrale will need this if I understand correctly)
  * Exchange list (combine types)
  * Create/edit exchange (combine types)

Again, need to see if we can make this happen with only admin creating exchange types.

**Comment by fosterlynn** [comment]
After the above, maybe minimal inventory.  Or may need it with either of the above.

* Resource list (Inventory) - this could be for everyone in view mode
* Create resource 
* Edit resource

**Comment by bum2** [comment]
The list "Minimal work planning" seems a good block of things to make a first update. Great!
About 'exchanges types' they should be set by admins only and show the common ones to the members.
About 'resource types' they will need the 'others' field and moderation system to let users suggest new ones and the the admins will review, order, reword, etc. Like the skills, they can be a common tree.
About 'resources' themselves, yes they need to see their related resources in the project editor (the members_agent.html page, when user is manager), and also create new resources of their project or project members. Perhaps not needed a 'inventory' page as such, if we can have a 'create new resource' button or a 'project resources' sub-page from the project editor.

**Comment by fosterlynn** [comment]
@bum2 thanks!  Mostly makes sense to me.  

But I'm thinking some more about the question of what to work on next in the project coordinator section.  Suggestion: we look at what projects are actually going to happen now, and tailor to them.  So Decentrale for example, may not need process logging or planning, I'm not sure.  If there are projects that are more to create something for a customer, then they will need it.

I won't start today, got other things that I have to do, so we have a little time.  I'll ask in the chat to see if we can have an idea of the first set of projects.

**Comment by fosterlynn** [comment]
@bum2 do you have thoughts about how to do the navigation for the project coordinators?  What do you think about a left nav bar on the project coordinator pages only?

(No rush, not a problem, I'll just work on individual pages.)

**Comment by fosterlynn** [comment]
I am running the app, and realizing that there is also the project-work page which needs to have the process tasks added to it.  Maybe I better do that first before planning.

**Comment by fosterlynn** [comment]
>I am running the app, and realizing that there is also the project-work page which needs to have the process tasks added to it. Maybe I better do that first before planning.

This is not as simple as replicating the Your Tasks section.  In the main app, the All Work page is the right functionality, except it would need to be limited to the projects the person is coordinator for.  

What do you think of using that format, or part of it?  (Choose an earlier start date to see actual data.)

**Comment by bum2** [comment]
Great Lynn, we can have this process tasks in the project-work yes. Thanks!
The navigation with a left sidebar is good for computers but very bad for mobiles, so i'm still using the label links as you have seen them as a member or project coordinator. Not tested in mobiles yet.
One day we will have time to do a good sub navigation system for all devices...

**Comment by bum2** [comment]
Lynn, please use the one that makes more sense for you... I can't focus on this yet.

**Comment by fosterlynn** [comment]
@bum don't worry, I'm on it.

**Comment by fosterlynn** [comment]
First piece of this done: project tasks in project-work.  Commented out a lot of links which are to features not there yet for project coordinators.  Still in project-planning branch, pushed.

**Comment by fosterlynn** [comment]
Next piece mostly done: project planning for project coordinators in the work app.  This includes:
* Viewing work orders (customer orders not implemented)
* Planning work (creating a new process and work order) - (but not from recipe)
* Viewing and changing order processes

It is not hooked up to the work app menu - but can be run from here: /work/order-list/.  Leaving it to @bum to decide where to hook it up.

It is pushed in the branch "project-planning".  Not merged to master.

And there is a bigger question for some time in the future - The project coordinator cannot create resource types or recipes, so the planning is harder.  Because to do that takes some training and understanding of the whole setup structure, since resource types and recipes go across projects right now.  They could ask a node admin to help with that.  This is perhaps an issue for discussion in the admin group some time.  For now, with Decentrale for example, it will be OK.

**Comment by fosterlynn** [comment]
^ Also, there is not a good way to plan a series of processes for manufacturing something without access to a recipe.  This is an issue in NRP too: https://github.com/valnet/valuenetwork/issues/444.  

Workflow plans have the capability, but they are simpler structures.

**Comment by fosterlynn** [comment]
Another piece pushed: exchanges (list and add/change/delete).  Still just in project-planning branch.

**Comment by bum2** [comment]
I'm working in your great project-planning branch @fosterlynn , mainly at nav level and style... 
See my commits in that branch, now at testocp.
I think some things are still no so clear:
- When in the work/agent/# page ('members_agent' or project editor page) they are listed the 'Active processes' of the project, but is not the same as the project open orders, listed in the other page. Every order implies a process? can we have project processes without a related project order? (i think a single order can have many processes), maybe we can show the project's orders also in the project page (open orders but also old finished orders).
- From the same 'members_agent' page (project page) will be good to have also a 'plan work' button that starts the planning config steps but with the context agent predefined to the project, so more simple steps and less fields.
- We really need to separate types (of exchange, of resource, patterns) in two groups: 'common' and 'only admins' (perhaps just adding a boolean field like 'public?'). Some of this types must be shown only if the user is a site admin, like the ones for FreedomCoop membership, processes, patterns, shares, fees, etc.

**Comment by fosterlynn** [comment]
@bum2 

>Every order implies a process? can we have project processes without a related project order?

Yes, if you create a process without specifying an order, it will create a new work order (at least it used to).

>maybe we can show the project's orders also in the project page (open orders but also old finished orders

makes sense

>From the same 'members_agent' page (project page) will be good to have also a 'plan work' button that starts the planning config steps but with the context agent predefined to the project, so more simple steps and less fields.

yes

>We really need to separate types (of exchange, of resource, patterns) in two groups: 'common' and 'only admins' (perhaps just adding a boolean field like 'public?'). Some of this types must be shown only if the user is a site admin, like the ones for FreedomCoop membership, processes, patterns, shares, fees, etc.

There was discussion of a feature related to this in the last assembly, and a team was formed, but hasn't gotten started yet (invoice numbers were higher priority).  I agree, we need this badly for collective groups to get going.  The feature was to allow projects  (actually any context agent) to create their own resource types, exchange types, and patterns, plus allow for global ones.  Then Freedom Coop as a context agent could have its own.  Will this cover your thoughts on "common" and "only admins"?

**Comment by fosterlynn** [comment]
In general, I like the direction you are going with the navigation -  that many things  (in this case planning work) will be already at the project (context) level, not have project selection on them.

**Comment by bum2** [comment]
> Then Freedom Coop as a context agent could have its own. Will this cover your thoughts on "common" and "only admins"?

Yes, general-common groups will have general-common types associated (usable for all child agents). Any project coordinator can set custom types after checking there isn't a suitable 'common' type for the case. The coordinators of general-common groups can update and add common types, and if a custom project's type is good to became a common one, only is needed to change the context of the type and set a general agent (either by a coop admin or by a general group coordinator). The 'only admins' internal types will be all set to another root group agent for internal FC projects, so they can share (or not) internal types (eg. membership share exchange type, etc).
I see there's a UseCase FK in the ExchangeType model, but seems not linked to any context agent. So, what will be better? to add a 'context_agent' field in the ExchangeType model or in the UseCase model? Maybe with the context hierachy approach for filtering, one day is not necessary the usecase? I have to study more this part of the code (and many others!) to understand good... 
Or perhaps some of you have some time and can do the addition of the FK field to link exchange Types to context agents (so I can focus in other pending stuff)? we can keep working in the 'project-planning' branch...

> In general, I like the direction you are going with the navigation - that many things (in this case planning work) will be already at the project (context) level, not have project selection on them.

Yes, it will be very nice to eliminate the step of choosing an agent, when going from the 'project' page, so it's more clear and the context is already set...

**Comment by fosterlynn** [comment]
>So, what will be better? to add a 'context_agent' field in the ExchangeType model or in the UseCase model?

We should add context agent to ExchangeType model.

I can at least assist in this work.

**Comment by bum2** [comment]
Ok, i will do it (with the migrations and so on) in the project-planning branch... 
Other small things to improve in this branch are: 
- Add a 'delete' button in the right of each item in the 'Order Items' list, in the work/order-plan page. So items can be added and also removed (only if more than one).
- The 'Change' button behaviour in every order in the work/order-list page should be as a modal box (not a new page).

If someone wants to do any of this other issues, feel free!
I've updated testocp with few UI changes in project-planning branch, and now i'll focus in the models...

**Comment by fosterlynn** [comment]
@bum2 nice.  I will let you know when I get some time too, but it is unclear.

So, for model changes, we would want to add the context_agent (EconomicAgent FK) to ExchangeType, ProcessPattern, and ResourceType.  And it should be optional.  Does this seem right?

**Comment by bum2** [comment]
Oook, now i have integrated the context-agent in the ExchangeType and the ResourceType models, and the related behaviors and filterings for the exchanges_all page and the exchange_logging page. Is pending to review where more the code needs to be adapted to behave 'by context'.
Next step is the ProcessPattern as you said, i want to focus this as soon as the exchanges part is working. Perhaps can be published a release still without the process planning part?, just to have more time to do it good...

**Comment by fosterlynn** [comment]
>Perhaps can be published a release still without the process planning part?,

I think for the hackathon OCP training, we could get away with the process side as it is.  We could just set up demo data ahead of time so that it isn't confusing.  So yes, I completely agree with publishing this release without the process side.  More and smaller releases would help a lot in general.
